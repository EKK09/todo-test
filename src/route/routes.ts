import { lazy } from 'react';
import { RouteProps } from 'react-router-dom';

export interface AuthRouteProps extends RouteProps {
    requireAuth?: boolean;
}

const routes: AuthRouteProps[] = [
  {
    path: '/',
    exact: true,
    requireAuth: true,
    component: lazy(() => import('src/page/HomePage')),
  },
  {
    path: '/login',
    component: lazy(() => import('src/page/LoginPage')),
  },
  {
    path: '*',
    component: lazy(() => import('src/page/NotFoundPage')),
  },
];

export default routes;
