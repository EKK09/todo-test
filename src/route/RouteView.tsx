import React from 'react';
import {
  Switch,
  Route,
  Redirect,
  useLocation,
} from 'react-router-dom';
import useUserStore from 'src/store/user';
import routes, { AuthRouteProps, ROUTE_PATH } from './routes';

const PrivateRoute = (props: AuthRouteProps) => {
  const userId = useUserStore((state) => state.id);
  const location = useLocation();

  if (userId) {
    return (
      <Route {...props} />
    );
  }
  return (
    <Redirect
      to={{
        pathname: ROUTE_PATH.LOGIN,
        state: { from: location },
      }}
    />
  );
};

const RouteView = ():React.ReactElement => (
  <Switch>
    {routes.map(
      (route, index) => (route.requireAuth
        ? <PrivateRoute key={index} {...route} /> : <Route key={index} {...route} />),
    )}
  </Switch>
);

export default RouteView;
