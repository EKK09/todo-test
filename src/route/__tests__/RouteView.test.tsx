import { render, waitFor, screen } from '@testing-library/react';
import React, { Suspense } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import useUserStore from 'src/store/user';
import RouteView from '../RouteView';
import '@testing-library/jest-dom';
import routes from '../routes';

jest.mock('src/page/LoginPage', () => () => 'Mock-LoginPage');
jest.mock('src/page/HomePage', () => () => 'Mock-HomePage');
jest.mock('src/page/NotFoundPage', () => () => 'Mock-NotFoundPage');

describe('RouteView ', () => {
  const fooFallback = <div>fooFallback</div>;
  it('show LoginPage', async () => {
    const history = createMemoryHistory();
    history.push('/login');
    const { container, getByText } = render(
      <Router history={history}>
        <Suspense fallback={fooFallback}>
          <RouteView />
        </Suspense>
      </Router>,
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
      expect(getByText('Mock-LoginPage')).toBeInTheDocument();
    });
  });

  describe('HomePage', () => {
    it('show LoginPage if user id has not been setted', async () => {
      useUserStore.setState({ id: null });

      const history = createMemoryHistory();
      history.push('/');
      const { container, getByText } = render(
        <Router history={history}>
          <Suspense fallback={fooFallback}>
            <RouteView />
          </Suspense>
        </Router>,
      );

      await waitFor(() => {
        expect(container).toMatchSnapshot();
        expect(getByText('Mock-LoginPage')).toBeInTheDocument();
      });
    });
    it('show HomePage if user id has been setted', async () => {
      useUserStore.setState({ id: 123 });

      const history = createMemoryHistory();
      history.push('/');
      const { container, getByText } = render(
        <Router history={history}>
          <Suspense fallback={fooFallback}>
            <RouteView />
          </Suspense>
        </Router>,
      );

      await waitFor(() => {
        expect(container).toMatchSnapshot();
        expect(getByText('Mock-HomePage')).toBeInTheDocument();
      });
    });
  });

  it('show NotFoundPage if no path matched', async () => {
    const fooPath = '/fooPath';
    const hasMatchPath: boolean = routes.some((route) => route.path === fooPath);

    expect(hasMatchPath).toBeFalsy();

    const history = createMemoryHistory();
    history.push(fooPath);
    const { container, getByText } = render(
      <Router history={history}>
        <Suspense fallback={fooFallback}>
          <RouteView />
        </Suspense>
      </Router>,
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
      expect(getByText('Mock-NotFoundPage')).toBeInTheDocument();
    });
  });
});
