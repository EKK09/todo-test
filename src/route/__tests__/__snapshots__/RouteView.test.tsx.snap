// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteView  HomePage show HomePage if user id has been setted 1`] = `
<div>
  <div>
    fooFallback
  </div>
</div>
`;

exports[`RouteView  HomePage show HomePage if user id has been setted 2`] = `
<div>
  Mock-HomePage
</div>
`;

exports[`RouteView  HomePage show LoginPage if user id has not been setted 1`] = `
<div>
  Mock-LoginPage
</div>
`;

exports[`RouteView  show LoginPage 1`] = `
<div>
  <div>
    fooFallback
  </div>
</div>
`;

exports[`RouteView  show LoginPage 2`] = `
<div>
  Mock-LoginPage
</div>
`;

exports[`RouteView  show NotFoundPage if no path matched 1`] = `
<div>
  <div>
    fooFallback
  </div>
</div>
`;

exports[`RouteView  show NotFoundPage if no path matched 2`] = `
<div>
  Mock-NotFoundPage
</div>
`;
